### ISSUE TITLE
Nesting Query Type

### ISSUE THREAD

**Issue by kayvaninvemo** [mmkal/slonik-tools#344]
Is there a way to generate the type for nesting queries?

```
    const q1 = sql`select * from my_table`;
    const q2 = sql`
      select q1.id, q1.my_col
      from (${q1}) q1
    `;
```

It gives me this error:

```
Query: "select q1.id, q1.my_col from ($1) a \\gdesc"
Result: "psql:<stdin>:1: ERROR:  syntax error at or near \"$1\"\nLINE 1: select q1.id, q1.my_col from ($1) a \n          
```

which the reason is quite obvious.

If it is inevitable, is there a way to tell the type generator not to try for that nesting query. I will generate the type manually for nesting queries, but if it gives error, it will be hard to manage the code. (It will cause us to ignore the errors)

```
    const q1 = sql<queries.MyTable>`select * from my_table`;
    const q2 = sql<my_queries.Q2>`
      select q1.id, q1.my_col
      from (${q1}) q1
    `;

  // this should be created by my self, but as I am using Pick<Type, Keys> it should be type safe enough
  export declare namespace my_queries {
    export type Q2 = Pick<queries.MyTable, "id" | "my_col">;
  }
  
  export declare namespace queries {
    // Generated by @slonik/typegen
  
    /** - query: `select * from my_table` */
    export interface MyTable {
      /** column: `public.my_table.id`, not null: `true`, regtype: `bigint` */
      id: number;
  
      /** column: `public.my_table.my_col`, not null: `true`, regtype: `text` */
      my_col: string;
  
      /** column: `public.my_table.my_col2`, not null: `true`, regtype: `text` */
      my_col2: string;
    }
  }
```

**Comment by mmkal** [comment]
You're right, types can't be generated for queries like this at the moment. I'm hoping that #315 will improve this somewhat. For now, see the readme "Limitations" section. You should find that since the tool doesn't know how to get a type for the query with a nested subquery, you can add a manual type and it won't be removed. But you will see an error message. To make the tool skip the query entirely, use a variable other than `sql`;

```ts
const q1 = sql`select * from my_table`;
const _sql = sql;
const q2 = _sql`
  select q1.id, q1.my_col
  from (${q1}) q1
`;
```

I'll leave this open until I add a case along these lines to the limitations section, but keep an eye on #315 for a fix that doesn't require manually adding a type.

**Comment by kayvaninvemo** [comment]
Thanks for the fast reply.

I want just to ask that, "vscode-sql-template-literal" does not do syntax highlighting when I use "_sql" is there a work around?

**Comment by mmkal** [comment]
You would have to ask the maintainer to support underscores too. Looks like it'd be as easy as adding `_` [here](https://github.com/ForbesLindesay/vscode-sql-template-literal/blob/bd3c4e6d3821413c60430e1342fd11b596747dea/syntaxes/lit-sql.json#L8). Feel free to point to this issue as an explanation. I think it's a reasonable request so hopefully would be accepted. If not, you could try [vscode-sql-template-literal](https://github.com/thebearingedge/vscode-sql-lit/blob/783aa0f3b3c5071800d4c9e0eb81a9c29b57626a/syntaxes/sql-lit.json#L13) which officially supports this library (see #269) so hopefully would be willing to make this small change.

**Comment by mmkal** [comment]
I'm also open to other ways of skipping codegen for certain queries like a comment directive if there are issues with the above suggestion for some reason, but I like `const _sql = sql` because

1. it's so intuitive why it works, it doesn't really require documentation outside of a "tips and tricks" section
2. There's no implementation or maintenance cost on the library side

**Comment by kayvaninvemo** [comment]
I used [SQL tagged template literals](https://marketplace.visualstudio.com/items?itemName=frigus02.vscode-sql-tagged-template-literals) and it works as expected.
```
    const sqlFragment = sql;
```

It also support typed sql fragments.
```
  sql<queries.MyTable>`select * from my_table`
```

You might consider change the readme to suggest this plugin instead
